File: src/epub_optimizer/epubcheck.py
"""Subprocess adapter for the official EPUBCheck CLI."""

from __future__ import annotations

import json
import math
import os
import shutil
import subprocess
from collections import defaultdict
from dataclasses import dataclass, field
from pathlib import Path


@dataclass(frozen=True)
class EpubCheckFinding:
    severity: str
    code: str
    message: str
    path: str | None = None
    line: int | None = None
    column: int | None = None

    @property
    def fingerprint(self) -> tuple[str, str, str]:
        return (self.code, " ".join(self.message.split()), self.path or "")


@dataclass(frozen=True)
class EpubCheckResult:
    available: bool
    status: str
    findings: list[EpubCheckFinding] = field(default_factory=list)
    returncode: int | None = None
    message: str | None = None
    occurrence_count: int = 0
    error_count: int = 0

    @property
    def errors(self) -> list[EpubCheckFinding]:
        return [f for f in self.findings if f.severity.lower() in {"error", "fatal"}]


@dataclass(frozen=True)
class EpubCheckComparison:
    input: EpubCheckResult
    output: EpubCheckResult
    persisting: list[EpubCheckFinding] = field(default_factory=list)
    resolved: list[EpubCheckFinding] = field(default_factory=list)
    introduced: list[EpubCheckFinding] = field(default_factory=list)

    @property
    def available(self) -> bool:
        return self.input.available and self.output.available


class EpubCheckRunner:
    def __init__(
        self,
        executable: str | None = None,
        jar: str | Path | None = None,
        java: str | None = None,
        timeout: float = 120.0,
    ) -> None:
        self.executable = executable or os.getenv("EPUBCHECK_EXECUTABLE")
        self.jar = (
            Path(jar or os.getenv("EPUBCHECK_JAR", ""))
            if (jar or os.getenv("EPUBCHECK_JAR"))
            else None
        )
        self.java = java or os.getenv("JAVA", "java")
        default_timeout = timeout if math.isfinite(timeout) and timeout > 0 else 120.0
        try:
            self.timeout = float(os.getenv("EPUBCHECK_TIMEOUT", default_timeout))
            if not math.isfinite(self.timeout) or self.timeout <= 0:
                self.timeout = default_timeout
        except (TypeError, ValueError):
            self.timeout = default_timeout

    def _command(self, epub: Path) -> list[str] | None:
        if self.executable:
            exe = shutil.which(self.executable) or self.executable
            if (
                not Path(exe).exists()
                and shutil.which(self.executable) is None
                and Path(self.executable).parent != Path(".")
            ):
                return None
            return [exe, str(epub), "--json", "-"]
        if self.jar and self.jar.is_file() and shutil.which(self.java):
            return [self.java, "-jar", str(self.jar), str(epub), "--json", "-"]
        return None

    def check(self, epub: Path) -> EpubCheckResult:
        command = self._command(epub)
        if command is None:
            return EpubCheckResult(
                False, "unavailable", message="EPUBCheck executable/JAR is unavailable."
            )
        try:
            completed = subprocess.run(
                command, capture_output=True, text=True, timeout=self.timeout, check=False
            )
        except (OSError, subprocess.TimeoutExpired) as exc:
            status = "timeout" if isinstance(exc, subprocess.TimeoutExpired) else "unavailable"
            return EpubCheckResult(False, status, message=str(exc))
        try:
            payload = json.loads(completed.stdout or completed.stderr or "{}")
        except json.JSONDecodeError:
            return EpubCheckResult(
                False,
                "error",
                returncode=completed.returncode,
                message="EPUBCheck returned invalid JSON.",
            )
        if completed.returncode not in {0, 1}:
            return EpubCheckResult(
                False,
                "error",
                returncode=completed.returncode,
                message="EPUBCheck failed to complete validation.",
            )
        if not isinstance(payload, dict) or not isinstance(payload.get("messages"), list):
            return EpubCheckResult(
                False,
                "error",
                returncode=completed.returncode,
                message="EPUBCheck returned an unexpected JSON report.",
            )
        raw = payload["messages"]
        findings = []
        occurrence_count = 0
        error_count = 0
        for item in raw if isinstance(raw, list) else []:
            if not isinstance(item, dict):
                continue
            raw_locations = item.get("locations")
            locations = raw_locations if isinstance(raw_locations, list) else [item]
            locations = [location for location in locations if isinstance(location, dict)]
            additional = item.get("additionalLocations", 0)
            additional_count = additional if isinstance(additional, int) else 0
            item_occurrences = len(locations) + max(additional_count, 0)
            occurrence_count += item_occurrences
            severity = str(item.get("severity", item.get("type", "error"))).lower()
            if severity in {"error", "fatal"}:
                error_count += item_occurrences
            for location in locations:
                findings.append(
                    EpubCheckFinding(
                        severity=severity,
                        code=str(item.get("ID", item.get("code", item.get("id", "unknown")))),
                        message=str(item.get("message", "")),
                        path=location.get("path") or location.get("file"),
                        line=location.get("line"),
                        column=location.get("column"),
                    )
                )
            findings.extend(
                EpubCheckFinding(
                    severity=severity,
                    code=str(item.get("ID", item.get("code", item.get("id", "unknown")))),
                    message=str(item.get("message", "")),
                )
                for _ in range(max(additional_count, 0))
            )
        return EpubCheckResult(
            True,
            "ok",
            findings,
            completed.returncode,
            None,
            occurrence_count,
            error_count,
        )


def compare_epubcheck(
    input_result: EpubCheckResult, output_result: EpubCheckResult
) -> EpubCheckComparison:
    before: dict[tuple[str, str, str], list[EpubCheckFinding]] = defaultdict(list)
    after: dict[tuple[str, str, str], list[EpubCheckFinding]] = defaultdict(list)
    for finding in input_result.errors:
        before[finding.fingerprint].append(finding)
    for finding in output_result.errors:
        after[finding.fingerprint].append(finding)
    persisting = []
    resolved = []
    introduced = []
    for fingerprint in before.keys() | after.keys():
        before_findings = before[fingerprint]
        after_findings = after[fingerprint]
        common = min(len(before_findings), len(after_findings))
        persisting.extend(after_findings[:common])
        resolved.extend(before_findings[common:])
        introduced.extend(after_findings[common:])

    # EPUBCheck caps the concrete locations it emits and summarizes the rest as
    # ``additionalLocations``. Archive changes can alter which occurrences make
    # that cap, so a pathless occurrence must be allowed to match a concrete
    # occurrence of the same diagnostic on the other side.
    def diagnostic(finding: EpubCheckFinding) -> tuple[str, str]:
        return (finding.code, " ".join(finding.message.split()))

    remaining_resolved = []
    introduced_by_diagnostic: dict[tuple[str, str], list[EpubCheckFinding]] = defaultdict(
        list
    )
    for finding in introduced:
        introduced_by_diagnostic[diagnostic(finding)].append(finding)
    for finding in resolved:
        candidates = introduced_by_diagnostic[diagnostic(finding)]
        match = next(
            (
                index
                for index, candidate in enumerate(candidates)
                if finding.path is None or candidate.path is None
            ),
            None,
        )
        if match is None:
            remaining_resolved.append(finding)
            continue
        persisting.append(candidates.pop(match))
    remaining_introduced = [
        finding
        for candidates in introduced_by_diagnostic.values()
        for finding in candidates
    ]
    baseline_by_diagnostic: dict[tuple[str, str], list[EpubCheckFinding]] = defaultdict(list)
    for finding in input_result.errors:
        baseline_by_diagnostic[diagnostic(finding)].append(finding)
    genuinely_introduced = []
    for finding in remaining_introduced:
        baseline = baseline_by_diagnostic[diagnostic(finding)]
        if finding.path is None:
            known_location = any(candidate.path is None for candidate in baseline)
        else:
            known_location = any(
                candidate.path is None or candidate.path == finding.path
                for candidate in baseline
            )
        if known_location:
            # EPUBCheck's capped ``additionalLocations`` count can change when
            # equivalent XHTML is reserialized. Once the same diagnostic and
            # resource exist in the baseline, multiplicity growth is advisory.
            persisting.append(finding)
        else:
            genuinely_introduced.append(finding)
    return EpubCheckComparison(
        input_result,
        output_result,
        persisting=persisting,
        resolved=remaining_resolved,
        introduced=genuinely_introduced,
    )

File: src/epubcheck.rs
use crate::process::{self, Interrupted, TimedOut};
use anyhow::Result;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::{
    collections::BTreeMap,
    env,
    path::{Path, PathBuf},
    process::Command,
    time::Duration,
};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Finding {
    pub severity: String,
    pub code: String,
    pub message: String,
    pub path: Option<String>,
    pub line: Option<i64>,
    pub column: Option<i64>,
}
impl Finding {
    pub fn is_error(&self) -> bool {
        matches!(self.severity.to_lowercase().as_str(), "error" | "fatal")
    }
    fn diagnostic(&self) -> (String, String) {
        (
            self.code.clone(),
            self.message
                .split_whitespace()
                .collect::<Vec<_>>()
                .join(" "),
        )
    }
    fn fingerprint(&self) -> (String, String, String) {
        let (code, message) = self.diagnostic();
        (code, message, self.path.clone().unwrap_or_default())
    }
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CheckResult {
    pub available: bool,
    pub status: String,
    pub findings: Vec<Finding>,
    pub returncode: Option<i32>,
    pub message: Option<String>,
    pub occurrence_count: usize,
    pub error_count: usize,
}
impl CheckResult {
    pub fn unavailable(status: &str, message: &str) -> Self {
        Self {
            available: false,
            status: status.into(),
            findings: Vec::new(),
            returncode: None,
            message: Some(message.into()),
            occurrence_count: 0,
            error_count: 0,
        }
    }
    pub fn errors(&self) -> Vec<Finding> {
        self.findings
            .iter()
            .filter(|f| f.is_error())
            .cloned()
            .collect()
    }
}
#[derive(Debug)]
pub struct Comparison {
    pub available: bool,
    pub persisting: Vec<Finding>,
    pub resolved: Vec<Finding>,
    pub introduced: Vec<Finding>,
}
pub trait Checker {
    fn check(&self, epub: &Path) -> Result<CheckResult>;
}
#[derive(Default)]
pub struct Runner {
    pub executable: Option<String>,
    pub jar: Option<PathBuf>,
    pub java: String,
    pub timeout: f64,
}
impl Runner {
    pub fn from_env() -> Self {
        let timeout = env::var("EPUBCHECK_TIMEOUT")
            .ok()
            .and_then(|s| s.parse::<f64>().ok())
            .filter(|n| n.is_finite() && *n > 0.0)
            .unwrap_or(120.0);
        Self {
            executable: env::var("EPUBCHECK_EXECUTABLE")
                .ok()
                .filter(|s| !s.is_empty()),
            jar: env::var_os("EPUBCHECK_JAR")
                .filter(|s| !s.is_empty())
                .map(PathBuf::from),
            java: env::var("JAVA").unwrap_or("java".into()),
            timeout,
        }
    }
}
impl Checker for Runner {
    fn check(&self, epub: &Path) -> Result<CheckResult> {
        let mut command = if let Some(exe) = &self.executable {
            Command::new(exe)
        } else if let Some(jar) = &self.jar
            && jar.is_file()
        {
            let mut c = Command::new(&self.java);
            c.arg("-jar").arg(jar);
            c
        } else {
            return Ok(CheckResult::unavailable(
                "unavailable",
                "EPUBCheck executable/JAR is unavailable.",
            ));
        };
        command.arg(epub).args(["--json", "-"]);
        let timeout = if self.timeout.is_finite() && self.timeout > 0.0 {
            self.timeout
        } else {
            120.0
        };
        let output = match process::capture(
            &mut command,
            None,
            false,
            Duration::try_from_secs_f64(timeout).ok(),
        ) {
            Ok(output) => output,
            Err(error) if error.is::<Interrupted>() => return Err(error),
            Err(error) => {
                return Ok(CheckResult::unavailable(
                    if error.is::<TimedOut>() {
                        "timeout"
                    } else {
                        "unavailable"
                    },
                    &error.to_string(),
                ));
            }
        };
        let json = if !output.stdout.is_empty() {
            &output.stdout
        } else if !output.stderr.is_empty() {
            &output.stderr
        } else {
            "{}"
        };
        Ok(parse_report(json, output.status.code().unwrap_or(-1)))
    }
}
pub fn parse_report(json: &str, status: i32) -> CheckResult {
    let error = |message: &str| {
        let mut result = CheckResult::unavailable("error", message);
        result.returncode = Some(status);
        result
    };
    let Ok(payload) = serde_json::from_str::<Value>(json) else {
        return error("EPUBCheck returned invalid JSON.");
    };
    if !matches!(status, 0 | 1) {
        return error("EPUBCheck failed to complete validation.");
    }
    let Some(messages) = payload.get("messages").and_then(Value::as_array) else {
        return error("EPUBCheck returned an unexpected JSON report.");
    };
    let mut result = CheckResult {
        available: true,
        status: "ok".into(),
        findings: Vec::new(),
        returncode: Some(status),
        message: None,
        occurrence_count: 0,
        error_count: 0,
    };
    for item in messages.iter().filter(|i| i.is_object()) {
        let fallback = vec![item.clone()];
        let locations = item
            .get("locations")
            .and_then(Value::as_array)
            .unwrap_or(&fallback);
        let extra = item
            .get("additionalLocations")
            .and_then(Value::as_i64)
            .unwrap_or(0)
            .max(0) as usize;
        let severity = item
            .get("severity")
            .or_else(|| item.get("type"))
            .and_then(Value::as_str)
            .unwrap_or("error")
            .to_lowercase();
        let code = item
            .get("ID")
            .or_else(|| item.get("code"))
            .or_else(|| item.get("id"))
            .and_then(Value::as_str)
            .unwrap_or("unknown")
            .to_string();
        let message = item
            .get("message")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string();
        for location in locations
            .iter()
            .filter(|l| l.is_object())
            .map(Some)
            .chain(std::iter::repeat_n(None, extra))
        {
            let finding = Finding {
                severity: severity.clone(),
                code: code.clone(),
                message: message.clone(),
                path: location
                    .and_then(|l| {
                        l.get("path")
                            .filter(|v| v.as_str().is_some_and(|s| !s.is_empty()))
                            .or_else(|| l.get("file"))
                    })
                    .and_then(Value::as_str)
                    .map(str::to_owned),
                line: location.and_then(|l| l.get("line")).and_then(Value::as_i64),
                column: location
                    .and_then(|l| l.get("column"))
                    .and_then(Value::as_i64),
            };
            result.occurrence_count += 1;
            result.error_count += usize::from(finding.is_error());
            result.findings.push(finding);
        }
    }
    result
}
pub fn compare(input: &CheckResult, output: &CheckResult) -> Comparison {
    let mut before: BTreeMap<_, Vec<Finding>> = BTreeMap::new();
    let mut after: BTreeMap<_, Vec<Finding>> = BTreeMap::new();
    for f in input.errors() {
        before.entry(f.fingerprint()).or_default().push(f);
    }
    for f in output.errors() {
        after.entry(f.fingerprint()).or_default().push(f);
    }
    let mut result = Comparison {
        available: input.available && output.available,
        persisting: Vec::new(),
        resolved: Vec::new(),
        introduced: Vec::new(),
    };
    for (key, findings) in &before {
        let matches = after.remove(key).unwrap_or_default();
        let common = findings.len().min(matches.len());
        result.persisting.extend_from_slice(&matches[..common]);
        result.resolved.extend_from_slice(&findings[common..]);
        result.introduced.extend_from_slice(&matches[common..]);
    }
    result.introduced.extend(after.into_values().flatten());
    result.resolved.retain(|f| {
        if let Some(index) = result.introduced.iter().position(|c| {
            c.diagnostic() == f.diagnostic() && (f.path.is_none() || c.path.is_none())
        }) {
            result.persisting.push(result.introduced.remove(index));
            false
        } else {
            true
        }
    });
    result.introduced.retain(|f| {
        let known = input.errors().iter().any(|c| {
            c.diagnostic() == f.diagnostic()
                && if f.path.is_none() {
                    c.path.is_none()
                } else {
                    c.path.is_none() || c.path == f.path
                }
        });
        if known {
            result.persisting.push(f.clone());
            false
        } else {
            true
        }
    });
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;
    fn report(path: Option<&str>, extra: usize) -> CheckResult {
        parse_report(&json!({"messages": [{"ID": "RSC-005", "message": "Bad structure", "severity": "ERROR", "locations": [{"path": path}], "additionalLocations": extra}]}).to_string(), 1)
    }
    #[test]
    fn comparison_preserves_opaque_locations_but_rejects_new_resources() {
        let a = report(Some("a.xhtml"), 0);
        let b = report(Some("b.xhtml"), 0);
        assert_eq!(compare(&a, &b).introduced.len(), 1);
        let a = report(Some("a.xhtml"), 3);
        assert!(compare(&a, &b).introduced.is_empty());
        assert_eq!(compare(&a, &b).resolved.len(), 3);
        let b = report(Some("a.xhtml"), 9);
        assert!(compare(&a, &b).introduced.is_empty());
        assert_eq!(b.error_count, 10);
    }
    #[test]
    fn malformed_checker_reports_are_unavailable() {
        for s in ["", "[]", "{}", "{\"messages\":{}}"] {
            assert!(!parse_report(s, 0).available);
        }
        assert!(!parse_report("{\"messages\":[]}", 2).available);
    }
}
